### PR TITLE
Band structure and density of states component overhaul and test app

### PR DESCRIPTION
**Pull request by munrojm**
- BSDOS component now also allows passing band structure and dos objects directly. 
- Plotting of the band structure traces has been sped up through an imposed energy window.
- Drop downs included to choose path conventions, projection types, and plot labels.
- Example app added with test.

### CODE REVIEW

**Comment by mkhorton** [comment]
Fantastic! Many thanks @munrojm 

FYI, this pattern:

```
# tell Crystal Toolkit about the app
ctc.register_app(app)

# to ensure all necessary components are loaded into layout
app.layout = ctc.crystal_toolkit_layout(my_layout)
```

has now been simplified:

```
ctc.register_crystal_toolkit(app, layout=my_layout, cache=None)  # cache optional
```

The old method still works, but this is just one less line to explain to people.

**Comment by mkhorton** [comment]
@munrojm it may help to explicitly pass an `id` to your component, I suspect this is related to a known issue related to how default ids are generated

File: crystal_toolkit/apps/examples/bs.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+
+mpid = {"mpid": "mp-149"}
+
+with MPRester() as m:
+    bandstructure_symm_line = m.get_bandstructure_by_material_id(mpid["mpid"])
+    density_of_states = m.get_dos_by_material_id(mpid["mpid"])
+
+# create the Crystal Toolkit component
+structure_component = ctc.BandstructureAndDosComponent(
```
Comment (mkhorton) [inline crystal_toolkit/apps/examples/bs.py]: e.g. put `ctcBandstructureAndDosComponent(..., id=)` here and reference that id in callbacks (also rename the variable so it's not `structure_component` anymore)

Comment (munrojm) [reply crystal_toolkit/apps/examples/bs.py]: Thanks! It is working now on my end.

**Comment by munrojm** [comment]
Ah, it's failing here now because I pull a band structure and dos with MPRester and it wants an API key to do so.

**Comment by mkhorton** [comment]
Ok, I have an API key available in the tests, I just need to enable it

**Comment by mkhorton** [comment]
Ok, should be good now. Not sure how to re-run unless you push another commit. You could also just commit a band structure and dos file if preferred.

**Pull request merged**